asciidoc/modules/con_mg_improvements-in-vertx-http-client.adoc

```diff
@@ -0,0 +1,203 @@
+
+* Sending the request headers
+* HTTP/2 specific operations such as setting a push handler, setting stream priority, pings, and so on.
+* Creating a NetSocket tunnel
+* Providing fine grained write control
+* Resetting a stream
+* Handling 100 continue headers manually
+
+The following example shows you how to create an `HTTPClientRequest`.
+
+[source,java,options="nowrap",subs="attributes+"]
+----
+client.request(HttpMethod.GET, 8080, "example.com", "/resource", ar -> {
+  if (ar.succeeded()) {
+    HttpClientRequest request = ar.result();
+    request.putHeader("content-type", "application/json")
+    request.send(new JsonObject().put("hello", "world"))
+      .onSuccess(response -> {
+      //
+      }).onFailure(err -> {
+      //
+       });
+     }
+})
+----
+
+== Removed the connection handler method from HTTP client request
+
+The `HttpClientRequest.connectionHandler()` method has been removed. Use `HttpClient.connectionHandler()` method instead to call connection handlers for client requests in your application.
+
+The following example shows how the `HttpClientRequest.connectionHandler()` method was used in {VertX} {3x} releases.
+
+[source,java,options="nowrap",subs="attributes+"]
+----
+client.request().connectionHandler(conn -> {
+  // Do something with connection
+}).end();
+----
+
+The following example shows you how to use the new `HttpClient.connectionHandler()` method.
+
+[source,java,options="nowrap",subs="attributes+"]
+----
+client.connectionHandler(conn -> {
+// Connection related code
+});
+----
+
+== HTTP client tunneling using the net socket method
+
+HTTP tunnels can be created using the `HttpClientResponse.netSocket()` method. In {VertX} {4} this method has been updated.
+
+To get a net socket for the connection of the request, send a socket handler in the request. The handler is called when the HTTP response header is received. The socket is ready for tunneling and can send and receive buffers.
+
+The following example shows how to get net socket for a connection in {VertX} {3x} releases.
+
+[source,java,options="nowrap",subs="attributes+"]
+----
+client.request(HttpMethod.CONNECT, uri, ar -> {
+  if (ar.succeeded()) {
+    HttpClientResponse response = ar.result();
+    if (response.statusCode() == 200) {
+      NetSocket so = response.netSocket();
+   }
+  }
+}).end();
+----
+
+The following example shows how to get net socket for a connection in {VertX} {4}:
+
+[source,java,options="nowrap",subs="attributes+"]
+----
+client.request(HttpMethod.CONNECT, uri, ar -> {
+}).netSocket(ar -> {
+  if (ar.succeeded()) {
+   // Got a response with a 200 status code
+   NetSocket so = ar.result();
+   // Go for tunneling
+  }
+}).end();
+----
+
+== New `send()` method in `HttpClient` class
+
+A new `send()` method is available in the `HttpClient` class. This method enables you to send requests in a generic way.
+
+The following code shows how to send a request in {VertX} {4}.
```

Comment:
@vietj what does "generic way" mean?

Reply:
I think this paragraph can be dropped, it's an old change that was dropped since it was written in wiki.